src/treecons.ml

```diff
@@ -0,0 +1,131 @@
+(** Tree constructors: a simpler representation of terms to be used by a
+    decision tree. *)
+
+open Extra
+
+(** A constructor is the representation of a symbol along with the number of
+    arguments to which it is applied. *)
+type treecons =
+  | Symb of
+      { c_mod : string list
+      (** Module name where the symbol of the constructor is defined. *)
+      ; c_sym : string
+      (** Symbol of the constructor. *)
+      ; c_ari : int
+      (** Arity of the considered constructor.  A same symbol representation
+          may generate several constructors with different arities. *) }
+  | Abst
+  (** An abstraction*)
+  | Vari of string
+  (** A free variable with a name. *)
+
+(** [pp o c] prints tree constructor [c] to output channel [o]. *)
+let pp : treecons pp = fun oc -> function
+  | Abst    -> Format.fprintf oc "λ"
+  | Vari(s) -> Format.pp_print_string oc s
+  | Symb(t) -> Format.fprintf oc "%s<:%d" t.c_sym t.c_ari
+
+(** [compare c d] is a comparison function for constructors; more efficient
+    than the pervasive. *)
+let compare : treecons -> treecons -> int = fun ca cb ->
+  match ca, cb with
+  | Symb(a), Symb(b) ->
+      begin match Int.compare a.c_ari b.c_ari with
+      | 0 ->
+          begin match String.compare a.c_sym b.c_sym with
+          | 0 -> Pervasives.compare a.c_mod b.c_mod
+          | x -> x
+          end
+      | x -> x
+      end
+  | x, y                 -> Pervasives.compare x y
+
+(** [eq c d] returns true iff [c] and [d] are equal regarding
+    [compare]. *)
+let eq : treecons eq = fun a b -> compare a b = 0
+
+module type MapSig =
```

Comment:
Comment missing. Shouldn't this be in extra.ml?

Reply:
Perhaps, but it is not intended to be really general

Reply:
I see. Why not abstracting the type key?

Reply:
Because I've been lazy